### PR TITLE
Optional Trace import using ActiveJob

### PR DESCRIPTION
**Pull request by gravitystorm**
Coming out of the conversations in #281 and #1852, I've started working on a mechanism to import and delete traces using the job queue, and avoiding the need for the [external gpx importer](https://github.com/openstreetmap/gpx-import) completely.

My first aim is to provide people with an option to import traces in development, by providing a configuration option to enable it, and fixing the icon creation to work again. This is what I've done so far, and I'm putting it up here for initial comments. It doesn't have any tests yet, hence the WIP label. But I think it could be merged as-is, since it's off by default and the icon methods have been broken for years anyway.

Beyond this stage, I would see the plan (in no particular order) as roughly:
* Write some tests to make sure this all works (done)
* Get the animated gifs working (see https://github.com/dark-panda/gd2-ffij/issues/18)
* Get OSMF to set the option to false in the chef scripts, then flip the option to true in the defaults, so that the job queue works by default for new developers / other deployments
* Fix the importing to use [activerecord-import](https://github.com/zdennis/activerecord-import), to cut down the number of inserts per trace

OSMF can then switch off the external importer and use this one instead.

Initial thoughts/comments/etc welcome. Big thanks to @mmd-osm for investigating both gd2-ffij and activerecord-import

### CODE REVIEW

**Comment by gravitystorm** [comment]
I think this is ready for review/merging, now that the tests are written. There's obviously two major limitations with this code right now (the lack of the animated gif for the large picture, and the 1-sql-insert-per-tracepoint thing), but I'd rather see those fixed those in future PRs rather than doing everything in one PR.

File: test/jobs/trace_importer_job_test.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    gpx = Minitest::Mock.new
+    def gpx.actual_points
+      5
+    end
+
+    trace.stub(:import, gpx) do
+      perform_enqueued_jobs do
+        TraceImporterJob.perform_now(trace)
+      end
+    end
```
Comment (tomhughes) [inline test/jobs/trace_importer_job_test.rb]: I'm struggling to understand this, or to find any documentation of the `stub` method that would help me. I assume it's a FactoryBot thing but although I find occasional references to it on google the actual documentation never seems to mention it.

I assume it stubs the `import` method but I don't see when this test would ever call that method? or why calling import which would normally be called from the job should actually run the job?

I'm probably just being dense but I'd like to understand what's going on ;-)

File: app/jobs/trace_destroyer_job.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class TraceDestroyerJob < ApplicationJob
+  queue_as :default
+
+  def perform(trace)
+    trace.destroy
+  rescue StandardError => ex
+    logger.info ex.to_s
+    ex.backtrace.each { |l| logger.info l }
```
[the remark this answers is not in the document]
Comment (gravitystorm) [reply app/jobs/trace_destroyer_job.rb]: Just copy+pasted from the previous daemon code. I agree, it makes sense to remove the exception handling from here.

File: test/jobs/trace_importer_job_test.rb (hunk 1 above)
Comment (gravitystorm) [reply test/jobs/trace_importer_job_test.rb]: It's from minitest/mock. 

https://www.rubydoc.info/gems/minitest/4.2.0/Object:stub

The stubbing takes a block, and the stub is only effective inside the block. So when we call `TraceImporterJob.perform_now(trace)`, the `trace.import` method has been overwritten to just return the (fake) gpx object and avoid doing anything else (parsing xml, creating images on disk etc).

I did all this before realising that I needed the FakeFS for other stuff anyway. So if you prefer, I could rework this to create a fully working `trace` object that actually runs `trace.import` for real, with suitable setup to put a valid xml file in the right place and clean up the side-effects of running the `.import` code. Or we can stick with mocking/stubbing, which is probably a better way to do it, but perhaps a bit impenetrable?

**Comment by mmd-osm** [comment]
> There's obviously two major limitations with this code right now (the lack of the animated gif for the large picture, and the 1-sql-insert-per-tracepoint thing), but I'd rather see those fixed those in future PRs rather than doing everything in one PR.

Animated gif support may be a bit tricky, as the repo owner hasn't been all that responsive lately. Maybe we'd have to fork it, and run our version of it, and find someone to finialize that animated gif part? 

Bulk INSERT is quite crucial from a performance pov, otherwise this thing is barely usable. Given that we already have that code lying around (https://github.com/openstreetmap/openstreetmap-website/issues/1852#issuecomment-427116280), and @tomhughes already reviewed a few iterations, I don't see why we couldn't integrate this part right away.

**Comment by gravitystorm** [comment]
> Bulk INSERT is quite crucial from a performance pov, otherwise this thing is barely usable

I worked on it last week too, after I made the pull request, and it was [pretty straightforward](https://github.com/gravitystorm/openstreetmap-website/commit/74e1b98ea2701b722ebbb124e03f272848c51676). I noticed that the way the import method currently works involved multiple passes at the XML parsing, so I want to look at that too. But in general, I'm a fan of incremental pull requests, to make the reviews more straightforward.

I'll work on this further next week.

**Comment by mmd-osm** [comment]
Some folks on Stackoverflow reported an issue with gd2-ffij on MacOS: "Couldn't load the gd2 library". This may be some documentation issue in INSTALL.md. I asked them to create a new issue here, as I can't reproduce this locally.

**Pull request merged**